### ISSUE TITLE
UnsatisfiedLinkError: Native method not found: nativeBeginImplicit

### ISSUE THREAD

**Issue by datrixo** [realm/realm-java#1535]
Hi!

Next "popular" crash reported from our crash analytics tools for version 82.2(and previous) is:

Fatal Exception: java.lang.UnsatisfiedLinkError
Native method not found: io.realm.internal.SharedGroup.nativeBeginImplicit:(J)J

java.lang.UnsatisfiedLinkError: Native method not found: 
io.realm.internal.SharedGroup.nativeBeginImplicit:(J)J
       at io.realm.internal.SharedGroup.nativeBeginImplicit(SharedGroup.java)
       at io.realm.internal.SharedGroup.beginImplicitTransaction(SharedGroup.java:117)
       at io.realm.Realm.(Realm.java:207)
       at io.realm.Realm.createAndValidate(Realm.java:594)
       at io.realm.Realm.create(Realm.java:564)
       at io.realm.Realm.getInstance(Realm.java:537)

System versions: 4.4.3, 4.4.4
Devices: Motorola XT1033, XT1040, XT1069

We using Dagger2, but this should'nt imply any problems.

**Comment by kneth** [comment]
Realm has a large `.so` file which is loaded the first time you open a Realm file. Somehow, the `.so` file isn't loaded. We have never been able to reproduce it but it has been reported a couple of times now.

Do you have access to the complete log from one of the crashes? It might give us a hint where to look.

**Comment by datrixo** [comment]
This is full stack trace reported by Crashlytics:

```
java.lang.UnsatisfiedLinkError: Native method not found: io.realm.internal.SharedGroup.nativeBeginImplicit:(J)J
       at io.realm.internal.SharedGroup.nativeBeginImplicit(SharedGroup.java)
       at io.realm.internal.SharedGroup.beginImplicitTransaction(SharedGroup.java:117)
       at io.realm.Realm.(Realm.java:207)
       at io.realm.Realm.createAndValidate(Realm.java:594)
       at io.realm.Realm.create(Realm.java:564)
       at io.realm.Realm.getInstance(Realm.java:537)
       at io.realm.RealmFactory.createRealm(RealmFactory.java:21)
       at io.realm.RealmFactory.getInstance(RealmFactory.java:12)
       at com.someappname.v2.dagger.modules.RealmModule.provideRealm(RealmModule.java:17)
       at com.someappname.v2.dagger.modules.RealmModule_ProvideRealmFactory.get(RealmModule_ProvideRealmFactory.java:23)
       at com.someappname.v2.dagger.modules.RealmModule_ProvideRealmFactory.get(RealmModule_ProvideRealmFactory.java:9)
       at dagger.internal.ScopedProvider.get(ScopedProvider.java:46)
       at com.someappname.v2.dagger.DaggerBaseComponent.realm(DaggerBaseComponent.java:126)
       at com.someappname.v2.dagger.DaggerAppComponent$1.get(DaggerAppComponent.java:45)
       at com.someappname.v2.dagger.DaggerAppComponent$1.get(DaggerAppComponent.java:42)
       at com.someappname.v2.App_MembersInjector.injectMembers(App_MembersInjector.java:35)
       at com.someappname.v2.App_MembersInjector.injectMembers(App_MembersInjector.java:11)
       at com.someappname.v2.dagger.DaggerAppComponent.inject(DaggerAppComponent.java:71)
       at com.someappname.v2.App.onCreate(App.java:89)
       at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1030)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4409)
       at android.app.ActivityThread.access$1500(ActivityThread.java:139)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1270)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:136)
       at android.app.ActivityThread.main(ActivityThread.java:5086)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
       at dalvik.system.NativeStart.main(NativeStart.java)
```

We have more weird crashes from realm.io (few occurences).
For example:
```
java.lang.UnsatisfiedLinkError: Native method not found: io.realm.internal.UncheckedRow.nativeGetString:(JJ)Ljava/lang/String;
       at io.realm.internal.UncheckedRow.nativeGetString(UncheckedRow.java)
       at io.realm.internal.UncheckedRow.getString(UncheckedRow.java:134)
```

Or:
```
java.lang.UnsatisfiedLinkError: Native method not found: io.realm.internal.Table.nativeWhere:(J)J
       at io.realm.internal.Table.nativeWhere(Table.java)
       at io.realm.internal.Table.where(Table.java:1413)
       at io.realm.RealmQuery.(RealmQuery.java:76)
       at io.realm.Realm.where(Realm.java:1213)
```

All of them for Motorola devices: XT1022, XT1032, XT1058 (Android 4.4.4)

Ps. Don't worry about io.realm.RealmFactory class.

**Comment by riyase** [comment]
on my Motorola motog-2 (dualSIm india)

FATAL EXCEPTION: main
java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/packagename-1/base.apk"],nativeLibraryDirectories=[/data/app/packagename-1/lib/arm, /vendor/lib, /system/lib]]] couldn't find "librealm-jni.so"
at java.lang.Runtime.loadLibrary(Runtime.java:366)
at java.lang.System.loadLibrary(System.java:989)
at io.realm.internal.RealmCore.loadLibrary(RealmCore.java:117)
at io.realm.internal.SharedGroup.<clinit>(SharedGroup.java:43)
at io.realm.internal.SharedGroupManager.<init>(SharedGroupManager.java:45)
at io.realm.BaseRealm.<init>(BaseRealm.java:87)
at io.realm.Realm.<init>(Realm.java:151)
at io.realm.Realm.createAndValidate(Realm.java:279)
at io.realm.Realm.create(Realm.java:247)
at io.realm.Realm.getInstance(Realm.java:219)
at io.realm.Realm.getInstance(Realm.java:182)

**Comment by riyase** [comment]
when i add 
classpath 'com.google.gms:google-services:1.4.0-beta3' in to "Project" gradle file
and
apply plugin: 'com.google.gms.google-services' in to "app" gradle file
Realm makes "UnsatisfiedLinkError"

i tested by removing above lines from gradle files.. then Realm working fine..

**Comment by erichkleung** [comment]
i'm having the same issues as above, and here's another along those same lines. once again, few crashes for each of these, but they keep popping up 

XT1031:
```
java.lang.UnsatisfiedLinkError: Native method not found: io.realm.internal.SharedGroup.nativeCloseReplication:(J)V
       at io.realm.internal.SharedGroup.nativeCloseReplication(SharedGroup.java)
       at io.realm.internal.SharedGroup.close(SharedGroup.java:172)
       at io.realm.Realm.close(Realm.java:192)
       at com.myapp.myapp.Application.checkForMigration(Application.java:134)
       at com.myapp.myapp.Application.onCreate(Application.java:91)
       at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1030)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4409)
       at android.app.ActivityThread.access$1500(ActivityThread.java:139)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1270)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:136)
       at android.app.ActivityThread.main(ActivityThread.java:5086)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
       at dalvik.system.NativeStart.main(NativeStart.java)
``` 

or here: 
DROID RAZR HD
```
java.lang.UnsatisfiedLinkError: Native method not found: io.realm.internal.Group.nativeGetTableNativePtr:(JLjava/lang/String;)J
       at io.realm.internal.Group.nativeGetTableNativePtr(Group.java)
       at io.realm.internal.Group.getTable(Group.java:229)
       at io.realm.Realm.getVersion(Realm.java:1332)
       at io.realm.Realm.initializeRealm(Realm.java:481)
       at io.realm.Realm.createAndValidate(Realm.java:448)
       at io.realm.Realm.create(Realm.java:418)
       at io.realm.Realm.getInstance(Realm.java:336)
       at io.realm.Realm.getInstance(Realm.java:298)
       at io.realm.Realm.getInstance(Realm.java:279)
       at com.myapp.myapp.MApplication.checkForMigration(MApplication.java:107)
       at com.myapp.myapp.MApplication.onCreate(MApplication.java:91)
       at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1030)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4425)
       at android.app.ActivityThread.access$1500(ActivityThread.java:139)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1270)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:136)
       at android.app.ActivityThread.main(ActivityThread.java:5102)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
       at dalvik.system.NativeStart.main(NativeStart.java)
```

**Comment by nhachicha** [comment]
@riyase did you check https://github.com/realm/realm-java/issues/1421#issuecomment-148616492 ? sounds like same issue

**Comment by riyase** [comment]
@nhachicha ya.. thanks.. :)

**Comment by kneth** [comment]
@riyase I close the issue as it sounds like you get it solved.

**Comment by datrixo** [comment]
Are you sure this should be closed? What is the final soulution? Because we didnt use 'com.google.gms:google-services:1.4.0-beta3' and this crash is still exist...

**Comment by dalinaum** [comment]
Do you still encounter this problem? @datrixo  I will reopen this.

**Comment by datrixo** [comment]
Yes. The answers above didnt have any connect with our stack of libs.
We are using: 
'com.android.tools.build:gradle:1.3.0'

and from google services:
'com.google.android.gms:play-services-analytics:7.0.0'

**Comment by riyase** [comment]
@kneth its not solved.. i want to use gcm in my app with realm

**Comment by kneth** [comment]
@riyase @datrixo can you create a small example to reproduce it?

**Comment by datrixo** [comment]
Nope. It's doesn't occurs on our machines. It's reported by crash reporting tool (many times). In previous comments I provided everything we know about this bug. If you need more specific informations about our app you can do it on my mail(available in profile) or just ask here. We really want help You to resolve this problem.

**Comment by datrixo** [comment]
Ok. We got next crashes, but now the message is little different:
```
java.lang.UnsatisfiedLinkError: 
**dlopen failed: "/mnt/asec/com.esky-1/lib/librealm-jni.so" is 32-bit instead of 64-bit**
       at java.lang.Runtime.loadLibrary(Runtime.java:371)
       at java.lang.System.loadLibrary(System.java:989)
       at io.realm.internal.RealmCore.loadLibrary(RealmCore.java:114)
       at io.realm.internal.SharedGroup.(SharedGroup.java:35)
       at io.realm.Realm.(Realm.java:205)
       at io.realm.Realm.createAndValidate(Realm.java:594)
       at io.realm.Realm.create(Realm.java:564)
       at io.realm.Realm.getInstance(Realm.java:537)
```

Device: HTC Desire 820

**Comment by kneth** [comment]
I believe that the HTC Desire 820 is an ARM 64, so the question is why the 32 bit version of `librealm-jni.so` is installed instead of the 64 bit version (both versions are distributed in the jar file). If you are using other libraries with only 32 bit `.so` files, you might see this. In #1640 a workaround is proposed (and used with success). It involves build Realm manually and excluding the ARM 64 build.

**Comment by LiuDeng** [comment]
this is still not resolve I  update to com.android.tools.build:gradle:1.4.0-beta6 and the apk build out have no librealm-jni.so under any folder of lib

**Comment by dalinaum** [comment]
Hi @LiuDeng. Did you check #1421?

**Comment by LiuDeng** [comment]
@dalinaum Thank you for your reply.I check that and i must to use gradle 1.4 for vector generator build in android studio 1.4. Now i fix that by just copy so to my project under jniLibs.

**Comment by beeender** [comment]
@datrixo @LiuDeng I think this problem has been fixed by gradle 1.5+  See https://realm.io/docs/java/latest/#couldnt-load-librealm-jniso for more details.

I am closing this issue now. Let me know if there is still a problem. Thanks!
